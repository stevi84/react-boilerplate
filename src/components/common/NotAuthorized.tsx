import Typography from '@mui/material/Typography';
import { IconButton, Paper, Stack } from '@mui/material';
import { useTranslation } from 'react-i18next';
import ArrowBack from '@mui/icons-material/ArrowBack';
import { useNavigate } from 'react-router-dom';

export const NotAuthorized = () => {
  const { t } = useTranslation();
  const navigate = useNavigate();

  return (
    <Paper sx={{ margin: 1 }}>
      <Stack direction={'column'} alignItems={'center'} justifyContent={'center'}>
        <Typography variant="h1" color="primary">
          403
        </Typography>
        <Typography variant="h4" color="textSecondary">
          Not Authorized
        </Typography>
        <Typography variant="body1" color="textSecondary">
          {t('403')}
        </Typography>
        <IconButton color="primary" onClick={() => navigate('/')}>
          <ArrowBack />
        </IconButton>
      </Stack>
    </Paper>
  );
};
