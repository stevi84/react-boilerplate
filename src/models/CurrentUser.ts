import { Components } from '../apis/api';

export type CurrentUser = Components.Schemas.CurrentUser;
export type Role = Components.Schemas.Role;
export const getEmptyCurrentUser = (): CurrentUser => ({
  loginName: '',
  firstName: '',
  surName: '',
  mail: '',
  roles: [],
});
