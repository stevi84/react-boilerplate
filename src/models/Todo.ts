import { Components } from '../apis/api';

export type Todo = Components.Schemas.Todo;

export const getEmptyTodo = (): Todo => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  owner: '',
  dueDate: '',
  description: '',
  completed: false,
});
