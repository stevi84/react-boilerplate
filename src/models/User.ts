import { Components } from '../apis/api';

export type User = Components.Schemas.User;

export const getEmptyUser = (): User => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  name: '',
  dateOfBirth: '',
  size: 0,
  weight: 0,
  email: '',
  phone: '',
});
