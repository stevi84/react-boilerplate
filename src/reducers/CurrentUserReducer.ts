import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from './Store';
import { CurrentUser, getEmptyCurrentUser } from '../models/CurrentUser';

export const currentUserInitialState: CurrentUser = getEmptyCurrentUser();

const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState: currentUserInitialState,
  reducers: {
    setCurrentUser(state, action: PayloadAction<CurrentUser>) {
      return action.payload;
    },
  },
});

const { actions, reducer } = currentUserSlice;
export const { setCurrentUser } = actions;
export { reducer as currentUserReducer };

export const currentUserSelector = (state: RootState): CurrentUser => state?.currentUser || currentUserInitialState;
