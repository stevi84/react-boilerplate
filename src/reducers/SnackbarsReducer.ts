import { PrepareAction, createAction, createReducer } from '@reduxjs/toolkit';
import { OptionsObject, SnackbarKey, SnackbarMessage } from 'notistack';
import { RootState } from './Store';

// https://stackoverflow.com/questions/62754584/how-can-i-use-notistick-any-snackbar-with-redux-toolkit-and-react
// https://codesandbox.io/s/github/iamhosseindhv/notistack/tree/master/examples/redux-example?file=/redux/actions.js:0-150

interface Snackbar {
  message: SnackbarMessage;
  options?: OptionsObject;
  key: SnackbarKey;
  dismissed: boolean;
}

interface SnackbarState {
  notifications: Snackbar[];
}

export const enqueueSnackbar = createAction<PrepareAction<{ notification: Snackbar }>>(
  'ENQUEUE_SNACKBAR',
  (message: SnackbarMessage, options?: OptionsObject) => ({
    payload: {
      notification: {
        message,
        options,
        key: (options && options.key) || new Date().getTime() + Math.random(),
        dismissed: false,
      },
    },
  })
);

export const closeSnackbar = createAction<PrepareAction<{ key?: SnackbarKey; dismissAll: boolean }>>(
  'CLOSE_SNACKBAR',
  (key?: SnackbarKey) => ({
    payload: {
      dismissAll: !key, // dismiss all if no key has been defined
      key,
    },
  })
);

export const removeSnackbar = createAction<PrepareAction<{ key: SnackbarKey }>>(
  'REMOVE_SNACKBAR',
  (key: SnackbarKey) => ({
    payload: {
      key,
    },
  })
);

export const snackbarsInitialState: SnackbarState = {
  notifications: [],
};

export const snackbarsReducer = createReducer(snackbarsInitialState, (builder) =>
  builder
    .addCase(enqueueSnackbar, (state, action) => ({
      notifications: [...state.notifications, { ...action.payload.notification }],
    }))
    .addCase(closeSnackbar, (state, action) => ({
      notifications: state.notifications.map((notification) =>
        action.payload.dismissAll || notification.key === action.payload.key
          ? { ...notification, dismissed: true }
          : { ...notification }
      ),
    }))
    .addCase(removeSnackbar, (state, action) => ({
      notifications: state.notifications.filter((n) => n.key !== action.payload.key),
    }))
);

export const notificationsSelector = (state: RootState): Snackbar[] => state?.snackbar?.notifications || [];
