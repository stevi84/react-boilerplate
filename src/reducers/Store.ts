import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { todosReducer } from './TodosReducer';
import { apiCallsReducer } from './ApiCallsReducer';
import { snackbarsReducer } from './SnackbarsReducer';
import { usersReducer } from './UsersReducer';
import { currentUserReducer } from './CurrentUserReducer';

export const store = configureStore({
  reducer: {
    snackbar: snackbarsReducer,
    todos: todosReducer,
    users: usersReducer,
    apiCalls: apiCallsReducer,
    currentUser: currentUserReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
