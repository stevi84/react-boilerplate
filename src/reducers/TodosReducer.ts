import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Todo } from '../models/Todo';
import { RootState } from './Store';

export const todosInitialState: Todo[] = [];

const todosSlice = createSlice({
  name: 'todos',
  initialState: todosInitialState,
  reducers: {
    setTodos(state, action: PayloadAction<Todo[]>) {
      return action.payload;
    },
    addTodo(state, action: PayloadAction<Todo>) {
      return state.concat(action.payload);
    },
    changeTodo(state, action: PayloadAction<Todo>) {
      return state.map((todo) => (todo.id !== action.payload.id ? todo : action.payload));
    },
    removeTodo(state, action: PayloadAction<number>) {
      return state.filter((todo) => todo.id !== action.payload);
    },
  },
});

const { actions, reducer } = todosSlice;
export const { setTodos, addTodo, changeTodo, removeTodo } = actions;
export { reducer as todosReducer };

export const todosSelector = (state: RootState): Todo[] => state?.todos || [];
