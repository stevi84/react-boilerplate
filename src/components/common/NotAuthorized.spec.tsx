import { describe, it, expect, vi } from 'vitest';
import { NotAuthorized } from './NotAuthorized';
import { fireEvent, screen } from '@testing-library/react';
import { currentUserEditor } from '../../test/data/CurrentUser';
import { renderWithProviders } from '../../test/Utils';
import { todo1 } from '../../test/data/Todo';
import { user1 } from '../../test/data/User';
import { RootState } from '../../reducers/Store';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

const { navigateMock } = vi.hoisted(() => ({ navigateMock: vi.fn() }));
vi.mock('react-router-dom', () => ({ useNavigate: () => navigateMock }));

vi.mock('../../hooks/UseNotifier', () => ({ useNotifier: vi.fn() }));

vi.mock('../../thunks/CurrentUserThunks', () => ({
  readCurrentUser: () => ({ type: 'readCurrentUser' }),
}));

const initialState: Partial<RootState> = {
  todos: [todo1],
  users: [user1],
  apiCalls: { runningReads: 0, runningSubmits: 0 },
  currentUser: currentUserEditor,
};

describe('NotAuthorized', () => {
  it('should equal saved snapshot', () => {
    const tree = renderWithProviders(<NotAuthorized />, { preloadedState: initialState }).asFragment();
    expect(tree).toMatchSnapshot();
  });

  it('should navigate to dashboard on button click', () => {
    renderWithProviders(<NotAuthorized />, { preloadedState: initialState });
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('ArrowBackIcon'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/');
  });
});
