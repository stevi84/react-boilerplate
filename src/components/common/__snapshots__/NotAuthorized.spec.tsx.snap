// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotAuthorized > should equal saved snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-3v46u1-MuiPaper-root"
  >
    <div
      class="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 css-utprum-MuiTypography-root"
      >
        403
      </h1>
      <h4
        class="MuiTypography-root MuiTypography-h4 css-zc6l1b-MuiTypography-root"
      >
        Not Authorized
      </h4>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        403
      </p>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
